cdap-metrics-sidecar/src/main/java/io/cdap/cdap/jmx/metrics/JMXMetricsCollector.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.jmx.metrics;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.google.inject.Inject;
+import com.sun.management.OperatingSystemMXBean;
+import io.cdap.cdap.api.Environment;
+import io.cdap.cdap.api.metrics.MetricsCollectionService;
+import io.cdap.cdap.api.metrics.MetricsCollector;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.proto.id.NamespaceId;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryMXBean;
+import java.lang.management.MemoryUsage;
+import java.lang.management.ThreadMXBean;
+import java.net.MalformedURLException;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.management.MBeanServerConnection;
+import javax.management.remote.JMXConnector;
+import javax.management.remote.JMXConnectorFactory;
+import javax.management.remote.JMXServiceURL;
+
+public class JMXMetricsCollector extends AbstractScheduledService {
+  private final CConfiguration cConf;
+  private static final Logger LOG = LoggerFactory.getLogger(JMXMetricsCollector.class);
+  private static final long MEGA_BYTE = 1024 * 1024;
+  private ScheduledExecutorService executor;
+  private final MetricsCollectionService metricsCollectionService;
+  private final String hostname;
+  private static final String SERVICE_URL_FORMAT = "service:jmx:rmi:///jndi/rmi://%s:%d/jmxrmi";
+  private final JMXServiceURL serviceUrl;
+  private  final Environment env;
+
+  @Inject
+  public JMXMetricsCollector(CConfiguration cConf, MetricsCollectionService metricsCollectionService, Environment env)
+    throws MalformedURLException {
+    this.cConf = cConf;
+    String serverUrl = String.format(SERVICE_URL_FORMAT, "localhost",
+                                     cConf.getInt(Constants.JMXMetricsCollector.SERVER_PORT));
+    this.metricsCollectionService = metricsCollectionService;
+    this.env = env;
+    this.hostname = env.getVariable("HOSTNAME");
```

Comment:
Since this service could be running in multiple pods, I added this to log statements to identify which pod this log came from. In k8s it's set automatically and is composed of pod name and container name.
I've removed it now.